### PR TITLE
fix Invalid mac updates

### PR DESCRIPTION
**Pull request by ganthern**
includes https://github.com/tutao/tutanota/pull/5980

### CODE REVIEW

File: src/calendar/model/CalendarModel.ts [hunk 1]
```diff
@@ -255,24 +254,45 @@ export class CalendarModel {
 	}
 
 	private async handleCalendarEventUpdate(update: CalendarEventUpdate): Promise<void> {
+		// we want to delete the CalendarEventUpdate after we are done, even, in some cases, if something went wrong.
+		let erase = true
```
Comment (charlag) [inline src/calendar/model/CalendarModel.ts:258]: I'm not a fan of this pre-initialized variable. I think we should either explicitly set it in each case (there are not so many of them) or better yet invoke some `this.eraseUpdate()` in each case.

File: src/desktop/DesktopNativeCryptoFacade.ts [hunk 2]
```diff
@@ -1,49 +1,69 @@
...
+	constructor(private readonly fs: FsExports, private readonly cryptoFns: CryptoFunctions, private readonly tfs: TempFs) {}
 
 	aesEncryptObject(encryptionKey: Aes256Key, object: number | string | boolean | ReadonlyArray<any> | {}): string {
 		const serializedObject = JSON.stringify(object)
 		const encryptedBytes = this.cryptoFns.aesEncrypt(encryptionKey, stringToUtf8Uint8Array(serializedObject))
 		return uint8ArrayToBase64(encryptedBytes)
 	}
 
-	aesDecryptObject(encryptionKey: Aes256Key, serializedObject: string): {} {
+	aesDecryptObject(encryptionKey: Aes256Key, serializedObject: string): Record<string, unknown> {
```
Comment (charlag) [inline src/desktop/DesktopNativeCryptoFacade.ts:27]: we assume that it is, in fact, an object but it doesn't have to be I think?

File: src/desktop/DesktopNativeCryptoFacade.ts [hunk 3]
```diff
@@ -1,49 +1,69 @@
...
 		const encData = await this.fs.promises.readFile(encryptedFileUri)
 		const bitKey = this.cryptoFns.bytesToKey(key)
 		const decData = this.cryptoFns.aesDecrypt(bitKey, encData, true)
 
 		const filesInDirectory = await this.fs.promises.readdir(targetDir)
-		const newFilename = nonClobberingFilename(filesInDirectory, path.basename(encryptedFileUri))
+		// since we're working purely in scratch space until putFileIntoDownloadsFolder
+		// is called, we could re-generate a random name here.
+		const writtenFileName = path.basename(encryptedFileUri)
+		const newFilename = nonClobberingFilename(filesInDirectory, writtenFileName)
```
Comment (charlag) [inline src/desktop/DesktopNativeCryptoFacade.ts:66]: with a random name we wouldn't need all this clobbering filenames I guess?

File: src/desktop/DesktopUtils.ts [hunk 4]
```diff
@@ -1,40 +1,41 @@
...
 
 	/**
-	 * open and close a file to make sure it exists
-	 * @param path: the file to touch
+	 * make sure we are allowed to take single-instance lock and clear up remaining tmp data
+	 * from previous runs if so.
 	 */
-	touch(path: string): void {
-		this.fs.closeSync(this.fs.openSync(path, "a"))
+	async cleanupOldInstance(): Promise<boolean> {
```
Comment (charlag) [inline src/desktop/DesktopUtils.ts:26]: needs a comment about return value

File: src/desktop/DesktopUtils.ts [hunk 5]
```diff
@@ -1,40 +1,41 @@
...
+		/**
+		 * doesn't clear tmp if:
+		 * * we're a second instance, the main instance may be using the tmp (we returned by now)
+		 * * there's a mailto link in the cli args, attachments may be located in the tmp
+		 * */
+		if (this.mailtoArg == null) this.tfs.clear()
+		return true
+	}
+
+	checkIsMailtoHandler(): Promise<boolean> {
```
Comment (charlag) [inline src/desktop/DesktopUtils.ts:37]: why make it async?

File: src/desktop/DesktopUtils.ts [hunk 6]
```diff
@@ -1,40 +1,41 @@
...
-	 * open and close a file to make sure it exists
-	 * @param path: the file to touch
+	 * make sure we are allowed to take single-instance lock and clear up remaining tmp data
+	 * from previous runs if so.
 	 */
-	touch(path: string): void {
-		this.fs.closeSync(this.fs.openSync(path, "a"))
+	async cleanupOldInstance(): Promise<boolean> {
+		if (!(await this.makeSingleInstance())) return false
+		/**
```
Comment (charlag) [inline src/desktop/DesktopUtils.ts:28]: this shouldn't be a jsdoc comment

File: src/desktop/DesktopUtils.ts [hunk 7]
```diff
@@ -101,50 +88,62 @@ export class DesktopUtils {
...
+			this.electron.app.quit()
+		}
+
+		return otherInstanceWillTerminate
+	}
 
-								return canStay
-							})
-			})
+	/** after we receive notification about another app instance being started, we need to decide
```
Comment (charlag) [inline src/desktop/DesktopUtils.ts:109]: sorry about being so annoying about the comments but jsdoc comment is either a single line or starts and ends with empty lines (`/**` and `*/`)

File: src/desktop/DesktopUtils.ts [hunk 8]
```diff
@@ -216,59 +196,18 @@ export class DesktopUtils {
...
-	}
-
 	getIconByName(iconName: string): NativeImage {
 		const iconPath = getResourcePath(`icons/${iconName}`)
 		return this.electron.nativeImage.createFromPath(iconPath)
 	}
+
+	async handleMailto(wm: WindowManager, mailToArg = this.mailtoArg) {
+		if (mailToArg) {
+			/*[filesUris, text, addresses, subject, mailToUrl]*/
```
Comment (charlag) [inline src/desktop/DesktopUtils.ts:206]: i would put one before each argument inline but this is also okay

File: src/desktop/files/DesktopFileFacade.ts [hunk 9]
```diff
@@ -138,49 +171,94 @@ export class DesktopDownloadManager {
...
-			await this.fs.promises.writeFile(savePath, file.data)
-		} catch (e) {
-			if (e.code === "EEXIST") {
-				// if the file exists we will try again and will pick another name
-				return this.writeDataFile(file)
-			} else {
-				throw e
-			}
+	async openFileChooser(boundingRect: IpcClientRect, filter: ReadonlyArray<string> | null): Promise<Array<string>> {
+		const opts: Record<string, unknown> = { properties: ["openFile", "multiSelections"] }
```
Comment (charlag) [inline src/desktop/files/DesktopFileFacade.ts:175]: I know it's not new but `opts` is actually `OpenDialogOptions` right? it just wants mutable filter for no reason but we can just give it a copy anyway

File: src/file/FileController.ts [hunk 10]
```diff
@@ -68,7 +68,8 @@ export abstract class FileController {
 				throw new ConnectionError("currently offline")
 			}
 		} finally {
-			// we don't necessarily know when the user is done with the temporary file that was opened.
+			// we don't necessarily know when the user is done with the temporary file that was opened
+			// because we don't have a handle to the opening application/process.
```
Comment (charlag) [inline src/file/FileController.ts:72]: but this does mean that we might nuke the opened file if we try to download another one

**Review by charlag — APPROVED** [review]
Refactoring with rearranging the structure/order of methods mixed with other changes makes it a bit hard to follow but looks good!

File: src/file/FileController.ts (hunk 10 above)
Comment (murilopereirame) [reply src/file/FileController.ts:72]: previously, that was handled by non clobbering file names. now, downloaded attachments inside the temp folder always have random names, so it should be okay

File: src/desktop/DesktopUtils.ts (hunk 7 above)
Comment (murilopereirame) [reply src/desktop/DesktopUtils.ts:109]: can´t we add this to eslint ~nig

File: src/desktop/DesktopNativeCryptoFacade.ts (hunk 3 above)
Comment (murilopereirame) [reply src/desktop/DesktopNativeCryptoFacade.ts:66]: we only fixed the problem for generated attachment uploads for now, this is also called in a context where we download actual attachments and don´t do randomization yet

File: src/desktop/DesktopUtils.ts (hunk 7 above)
Comment (ganthern) [reply src/desktop/DesktopUtils.ts]: there's https://github.com/hosseinmd/prettier-plugin-jsdoc but that requires prettier 3

Comment (charlag) [reply src/desktop/DesktopUtils.ts]: it doesn't?
https://github.com/hosseinmd/prettier-plugin-jsdoc?tab=readme-ov-file#supported-prettier-version

and also we could upgrade prettier anyway
